### DISCUSSION TITLE
Generate pattern/shortcut classes?

### DISCUSSION THREAD

**Discussion by bengry** [tailwindlabs/tailwindcss#11695] [Help]
We work in a large codebase that has some patterns that are often repeated. For example, a "default" border, which may be on the entire element, or one part of it (e.g. `border-top` only). This border can of course be in any one of the colors defined in the tailwind theme.

Currently to set a border top in tailwind we have to do this (`black` color used in this example):
```html
<div class="border-0 border-t-black border-t border-solid"></div>
```
We'd like to have some utilities for setting a border for every color, for every direction (i.e. top, bottom, left, right, all 4 sides):
```
.pattern_border-black // equivalent to `border-black border-solid border`
.pattern_border-t-red-50 // equivalent to `border-0 border-t-red-50 border-t border-solid`
...
```

Other patterns that repeat themselves are `.flex.flex-row` and `.flex.flex-col`, usually with a `gap` defined on them. So ideally I could also generate something like `.stack-4` which is equivalent to `.flex.flex-col.gap-4`.

I think this is usually called Patterns (in [PandaCSS](https://panda-css.com/docs/concepts/patterns)) or Shortcuts (in [UnoCSS](https://unocss.dev/config/shortcuts) & [Windi CSS](https://windicss.org/features/shortcuts.html#shortcuts)), but I'm sure there are other names for it.

Is this possible to do in Tailwind, including by authoring a plugin that auto-creates these for all border variants? I haven't been able to find something like this (besides [this](https://github.com/tailwindlabs/tailwindcss/discussions/9442#discussion-4430341), which seems like a subset of what I'm asking)

**Comment by wongjn** [comment]
If you are using preflight, you shouldn't need to specify `border-style: solid` as it should be the default. Furthermore, you shouldn't need to specify `border-0` either as again, `border-width: 0` should be the default with preflight enabled. Then, you can specify border color generally and only need the border side to be on the width class, like `border-t border-red-500`. This doesn't feel like this would need another level of abstraction, but if you would like it anyway (and for the flex combinations too), then consider [reading the documentation on adding utilities via a Tailwind plugin](https://tailwindcss.com/docs/plugins#adding-utilities).

**Reply by wongjn** [reply]
> We don't preflight (for legacy reasons, unfortunately). I read the docs, but couldn't find anything on adding values to things like borders - I want to add utilities for all `border-*` combinations basically, and all examples show how to do this for a list of values from the `theme`, not from something internal to Tailwind. In addition, these utilities set CSS values, not classes. i.e. there's no "`@apply`"-like syntax within the config, only access to the `theme` is possible.
> 
> Anyway, I think what I want is more akin to a "component" in tailwind, which I suppose `matchComponents` would help with, but there are no examples on how to use it, and it isn't exactly clear how it works or differs from `matchUtilities`.[^1]

> […] all examples show how to do this for a list of values from the theme, not from something internal to Tailwind.

You shouldn't need anything internal to Tailwind to achieve your goal.

> […] these utilities set CSS values, not classes […]

Indeed, you'd apply the CSS values as appropriate in the plugin-generated CSS rules.

> […] i.e. there's no `@apply`-like syntax within the config, only access to the theme is possible.

Incorrect. There is `@apply`-like syntax:

```js
addUtilities({
  '.foo': {
    '@apply border': {},
  },
});
```

> […] I suppose `matchComponents` would help with, but there are no examples on how to use it […]

You'd use `matchComponents()` like [`matchUtilities()`](https://tailwindcss.com/docs/plugins#dynamic-utilities).

> […] it isn't exactly clear how [`matchComponents()`] works or differs from matchUtilities.

You'd use `matchComponents()` like `matchUtilities()`. `matchComponents()` differs from `matchUtilities()` in that CSS rules generated by `matchComponents()` would:
- Be placed in the `components` `@layer`. `matchUtilities()` rules would be placed in the the `utilities` `@layer`.
- Not respect the  `important` configuration option by default. `matchUtilities()` rules would.

[^1]: https://github.com/tailwindlabs/tailwindcss/discussions/11695#discussioncomment-6563189

**Reply by bengry** [reply]
Thanks for the response (and sorry for taking so long to reply), this helped me out, and together with [this answer](https://github.com/tailwindlabs/tailwindcss/discussions/11695#discussioncomment-6583303) I was able to get 80% of the way to what I wanted.

**Comment by bengry** [comment]
> If you are using preflight, you shouldn't need to specify `border-style: solid` as it should be the default. Furthermore, you shouldn't need to specify `border-0` either as again, `border-width: 0` should be the default with preflight enabled. Then, you can specify border color generally and only need the border side to be on the width class, like `border-t border-red-500`. This doesn't feel like this would need another level of abstraction, but if you would like it anyway (and for the flex combinations too), then consider [reading the documentation on adding utilities via a Tailwind plugin](https://tailwindcss.com/docs/plugins#adding-utilities).

We don't preflight (for legacy reasons, unfortunately).
I read the docs, but couldn't find anything on adding values to things like borders - I want to add utilities for all `border-*` combinations basically, and all examples show how to do this for a list of values from the `theme`, not from something internal to Tailwind.
In addition, these utilities set CSS values, not classes. i.e. there's no "@apply"-like syntax within the config, only access to the `theme` is possible.

Anyway, I think what I want is more akin to a "component" in tailwind, which I suppose `matchComponents` would help with, but there are no examples on how to use it, and it isn't exactly clear how it works or differs from `matchUtilities`.

**Reply by wongjn** [reply]
Replied at https://github.com/tailwindlabs/tailwindcss/discussions/11695#discussioncomment-6563333 to keep the reply chain together.

**Comment by ony3000** [comment]
This is a bit complicated, but it allows you to make any set of border classes you want into a utility class (e.g. `pattern_border-black`, `pattern_border-t-red-500`, `pattern_border-y-blue-700/50`).

```javascript
// tailwind.config.js

const plugin = require('tailwindcss/plugin');

/** @type {import('tailwindcss').Config} */
module.exports = {
  theme: {
    // ...
  },
  plugins: [
    plugin(({ matchUtilities, theme }) => {
      [
        ['', { width: ['border-width'], color: ['border-color'] }],
        [
          '-x',
          {
            width: ['border-left-width', 'border-right-width'],
            color: ['border-left-color', 'border-right-color'],
          },
        ],
        [
          '-y',
          {
            width: ['border-top-width', 'border-bottom-width'],
            color: ['border-top-color', 'border-bottom-color'],
          },
        ],
        ['-t', { width: ['border-top-width'], color: ['border-top-color'] }],
        ['-r', { width: ['border-right-width'], color: ['border-right-color'] }],
        ['-b', { width: ['border-bottom-width'], color: ['border-bottom-color'] }],
        ['-l', { width: ['border-left-width'], color: ['border-left-color'] }],
      ].forEach(([directionSuffix, properties]) => {
        matchUtilities(
          {
            [`pattern_border${directionSuffix}`]: (hexColor, extra) => {
              const hexAlpha =
                extra.modifier === null
                  ? ''
                  : Math.round(Number(extra.modifier) * 255).toString(16);

              return {
                ...properties.width.reduce(
                  (prevRules, currentPropertyName) => ({
                    ...prevRules,
                    [currentPropertyName]: '1px',
                  }),
                  {},
                ),
                borderStyle: 'solid',
                ...properties.color.reduce(
                  (prevRules, currentPropertyName) => ({
                    ...prevRules,
                    [currentPropertyName]: `${hexColor}${hexAlpha}`,
                  }),
                  {},
                ),
              };
            },
          },
          {
            values: Object.entries(theme('colors') ?? {}).reduce(
              (prevResult, [colorName, value]) => ({
                ...prevResult,
                ...(typeof value === 'string'
                  ? { [colorName]: value }
                  : Object.entries(value).reduce(
                      (prevFlattenColors, [weight, hexColor]) => ({
                        ...prevFlattenColors,
                        [`${colorName}${weight === 'DEFAULT' ? '' : `-${weight}`}`]: hexColor,
                      }),
                      {},
                    )),
              }),
              {},
            ),
            modifiers: theme('opacity'),
          },
        );
      });
    }),
  ],
};
```

**Reply by bengry** [reply]
Thanks! I was able to modify this a bit to get what I wanted. Unfortunately, these don't expand at runtime, so I doubt something like [`tailwind-merge`](https://www.npmjs.com/package/tailwind-merge), but should be good enough.
